python/variable_init.cpp

```diff
@@ -0,0 +1,456 @@
+    }
+    if (with_variance.value_or(true) && !variances.is_none()) {
+      copy_array_into_view(cast_to_array_like<T>(variances, actual_unit),
+                           variable.template variances<T>(), dims);
+    }
+    return variable;
+  }
+};
+
+Variable make_variable_array(const py::object &dim_labels,
+                             const py::object &shape, const py::object &values,
+                             const py::object &variances,
+                             const units::Unit unit, DType dtype,
+                             const std::optional<bool> with_variance) {
+  using opt_array = std::optional<py::array>;
+  // Let numpy figure out conversions from lists, tuples, etc.
+  const opt_array values_array{values.is_none() ? opt_array{}
+                                                : py::array(values)};
+  const opt_array variances_array{variances.is_none() ? opt_array{}
+                                                      : py::array(variances)};
+  dtype = common_dtype(
+      values_array ? py::object(*values_array) : py::none(),
+      variances_array ? py::object(*variances_array) : py::none(), dtype, true);
+  const auto dims = build_dimensions(dim_labels.cast<std::vector<Dim>>(), shape,
+                                     values_array, variances_array);
+  return core::CallDType<
+      double, float, int64_t, int32_t, bool, scipp::core::time_point,
+      std::string, python::PyObject>::apply<MakeVariableArray>(dtype, dims,
+                                                               values,
+                                                               variances, unit,
+                                                               with_variance);
+}
+
+std::string make_scalar_array_conflict_message(const py::object &values,
+                                               const py::object &variances,
+                                               const py::object &value,
+                                               const py::object &variance) {
+  std::ostringstream oss;
+  oss << "Scalar and array arguments cannot both be used to initialize "
+         "variables. Got ";
+  bool first = true;
+  if (!values.is_none()) {
+    oss << "'values'";
+    first = false;
+  }
+  if (!variances.is_none()) {
+    if (!first)
+      oss << " and ";
+    oss << "'variances'";
+  }
+  oss << " which cannot be combined with ";
+  first = true;
+  if (!value.is_none()) {
+    oss << "'value'";
+    first = false;
+  }
+  if (!variance.is_none()) {
+    if (!first)
+      oss << " and ";
+    oss << "'variance'";
+  }
+  oss << '.';
+  return oss.str();
+}
+
+enum class ConstructorType { Array, Scalar, Default };
+
+ConstructorType
+identify_constructor(const py::object &dims, const py::object &shape,
+                     const py::object &values, const py::object &variances,
+                     const py::object &value, const py::object &variance) {
```

Comment:
Do we really need to `value` and `variance` args to `sc.Variable`? With the `sc.scalar` wrapper, can't we do everything using `dims`, `values`, and `variances`? Would that simplify things and make for a nicer signature of `sc.Variable`?

Reply:
Strictly speaking, no. We could inspect the types and shape of `values` and `variances` and the presence of `dims` to deduce whether it is a scalar. But this is not exactly trivial because
```.py
sc.Variable(values=[1, 2])  
sc.Variable(values=np.array(1))
```
are these errors or make scalars of PyObject?

We would need to specialise the `values` and `variances` casters because you cannot have pybind11 arrays of, e.g., `Variable`. So the current plain cast to `py::array` would not work.

See also https://github.com/scipp/scipp/pull/1994#discussion_r656940109

Reply:
Why not just make `dims` a required argument? Pass `dims=[]` to make a scalar, that's it.